deepspeed/runtime/pipe/engine.py

```diff
@@ -990,9 +990,7 @@ def _exec_send_grads(self, buffer_id):
             if isinstance(inputs, tuple):
                 first_input = inputs[0]
                 assert all([torch.is_tensor(elt) for elt in inputs[1:]])
-                inputs_grad_tail = [
-                    elt.grad for elt in inputs[1:] if elt.grad is not None
-                ]
+                inputs_grad_tail = [elt.grad for elt in inputs[1:]]
```

Comment:
It's been ages since I worked on this code unfortunately...could the extra `None`s result in mismatched number of communications in the backward pass from stage? I think that was the original intention. Our first version of pipelined GPT in Megatron used floating-point attention masks, but once switching to non-differentiable types things got trickier.

Reply:
We are seeing errors happen because of the lack of `None`s currently, as shown in https://github.com/EleutherAI/gpt-neox/issues/884